proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2948,198 +3035,256 @@ lemma cancelIPC_ccorres1:
+                         in ccorres_symb_exec_r_known_rv)
+                apply (rule conseqPre, vcg)
+                apply (clarsimp simp: st_tcb_at'_def)
+                apply (frule (1) obj_at_cslift_tcb)
+                apply (clarsimp simp: typ_heap_simps ctcb_relation_BlockedOnReceive_blockingObject)
+               apply ceqv
               apply (rule ccorres_symb_exec_r)
-                apply (simp only: fun_app_def simp_list_case_return
-                                  return_bind ccorres_seq_skip)
+                apply (simp only: list_case_If)
                 apply (rule ccorres_rhs_assoc2)
                 apply (rule ccorres_rhs_assoc2)
                 apply (rule ccorres_rhs_assoc2)
                 apply (ctac (no_vcg) add: cancelIPC_ccorres_helper)
-                  apply (ctac add: setThreadState_ccorres_valid_queues')
-                 apply (wp hoare_vcg_all_lift set_ep_valid_objs' | simp add: valid_tcb_state'_def split del: if_split)+
-                apply (simp add: ThreadState_defs)
+                  apply (rule_tac P="tcb_at' thread" in ccorres_cross_over_guard)
+                  apply (rule_tac xf'=ret__unsigned_longlong_'
+                              and val="thread_state_to_tsType threadState"
+                              and R="st_tcb_at' ((=) threadState) thread"
+                              and R'=UNIV
+                              in ccorres_symb_exec_r_known_rv)
+                     apply (rule conseqPre, vcg)
+                     apply (clarsimp simp: st_tcb_at'_def)
+                     apply (frule (1) obj_at_cslift_tcb)
+                     apply (fastforce simp: typ_heap_simps ctcb_relation_thread_state_to_tsType
+                                            thread_state_to_tsType_eq_BlockedOnReceive)
+                    apply ceqv
+                   apply ccorres_rewrite
+                   apply (rule ccorres_rhs_assoc)
+                   apply (rule ccorres_rhs_assoc)
+                   apply (rule_tac xf'=ret__unsigned_longlong_'
+                               and val="option_to_0 ro"
+                               and R="st_tcb_at' ((=) threadState) thread"
+                               and R'=UNIV
+                               in ccorres_symb_exec_r_known_rv)
+                      apply (rule conseqPre, vcg)
+                      apply (clarsimp simp: st_tcb_at'_def)
+                      apply (frule (1) obj_at_cslift_tcb)
+                      apply (clarsimp simp: typ_heap_simps
+                                            ctcb_relation_BlockedOnReceive_replyObject)
+                     apply ceqv
+                    apply csymbr
+                    apply (rule_tac P="ro \<noteq> Some 0" in ccorres_gen_asm)
+                    apply wpc
+                      \<comment> \<open>None\<close>
+                     apply simp
+                     apply (ctac add: setThreadState_ccorres)
+                    \<comment> \<open>Some\<close>
```

Comment:
for these None/Some comments (good idea), would be more useful to say what they refer to; this is obvious when you have the proof open, but not so obvious when reading the text, since it comes out of `wpc`